src/streamable.rs

```diff
@@ -0,0 +1,625 @@
+        for _ in 0..len {
+            ret.push(T::parse(input)?);
+        }
+        Ok(ret)
+    }
+}
+
+impl Streamable for String {
+    fn stream(&self, out: &mut Vec<u8>) -> Result<()> {
+        // bytes is the UTF-8 sequence
+        let bytes = self.bytes();
+        if bytes.len() > u32::MAX as usize {
+            Err(Error::InputTooLarge)
+        } else {
+            (bytes.len() as u32).stream(out)?;
+            out.extend(bytes);
+            Ok(())
+        }
+    }
+
+    fn parse(input: &mut InputStream) -> Result<Self> {
+        let len = u32::parse(input)?;
+        Ok(String::from(
+            std::str::from_utf8(input.read(len as usize)?).map_err(|_| Error::InvalidString)?,
+        ))
+    }
+}
+
+impl Streamable for bool {
+    fn stream(&self, out: &mut Vec<u8>) -> Result<()> {
+        out.extend_from_slice(if *self { &[1] } else { &[0] });
+        Ok(())
+    }
+    fn parse(input: &mut InputStream) -> Result<Self> {
+        let val = input.read(1)?[0];
+        match val {
+            0 => Ok(false),
+            1 => Ok(true),
+            _ => Err(Error::InvalidBool),
+        }
+    }
+}
+
+impl<T: Streamable> Streamable for Option<T> {
+    fn stream(&self, out: &mut Vec<u8>) -> Result<()> {
+        match self {
+            None => {
+                out.push(0);
+            }
+            Some(v) => {
+                out.push(1);
+                v.stream(out)?;
+            }
+        }
+        Ok(())
+    }
+    fn parse(input: &mut InputStream) -> Result<Self> {
+        let val = input.read(1)?[0];
+        match val {
+            0 => Ok(None),
+            1 => Ok(Some(T::parse(input)?)),
+            _ => Err(Error::InvalidOptional),
+        }
+    }
+}
+
+impl<T: Streamable, U: Streamable> Streamable for (T, U) {
+    fn stream(&self, out: &mut Vec<u8>) -> Result<()> {
+        self.0.stream(out)?;
+        self.1.stream(out)?;
+        Ok(())
+    }
+    fn parse(input: &mut InputStream) -> Result<Self> {
+        Ok((T::parse(input)?, U::parse(input)?))
+    }
+}
+
+impl<T: Streamable, U: Streamable, V: Streamable> Streamable for (T, U, V) {
+    fn stream(&self, out: &mut Vec<u8>) -> Result<()> {
+        self.0.stream(out)?;
+        self.1.stream(out)?;
+        self.2.stream(out)?;
+        Ok(())
+    }
+    fn parse(input: &mut InputStream) -> Result<Self> {
+        Ok((T::parse(input)?, U::parse(input)?, V::parse(input)?))
+    }
+}
+
+impl<T: Streamable, U: Streamable, V: Streamable, W: Streamable> Streamable for (T, U, V, W) {
+    fn stream(&self, out: &mut Vec<u8>) -> Result<()> {
+        self.0.stream(out)?;
+        self.1.stream(out)?;
+        self.2.stream(out)?;
+        self.3.stream(out)?;
+        Ok(())
+    }
+    fn parse(input: &mut InputStream) -> Result<Self> {
+        Ok((
+            T::parse(input)?,
+            U::parse(input)?,
+            V::parse(input)?,
+            W::parse(input)?,
+        ))
+    }
+}
+
+// ===== TESTS ====
```

Comment:
all of these tests were copied over from the `de.rs` and `ser.de`. But the `NotSupported` tests were removed (because those are compile-time errors now)